Add custom diagonalization tests

Yash-10 opened the pull request:
Issue #220

- I have checked that the tests pass locally with and without the optional libraries, except for `cupy` since its installation was a bit tricky initially (I am trying testing that locally). The workflow here might suggest whether it fails or not.
- Initially, I tried to parametrize the tests with all the available `esys` or `evals` methods but realized that some combinations may not be suitable naturally. This is why the methods are hardcoded inside the functions.
- A quick question: if the custom diagonalization method is a string, should providing an eval method to the `esys_method` argument or vice versa raise an explicit error? Currently, the error reads `ValueError: too many values to unpack (expected 2)` when an evals method is passed to `esys_method`.

(Installation of `cupy` suggested that using an old CUDA version (cuda90) gives an error `AttributeError: module 'cupy' has no attribute 'linalg'` when running the function `evals_cupy_dense`, which probably suggests that `cupy`'s minimum version may need to be pinned. Please let me know if this needs to be a separate issue, and I'll do that!)

petergthatsme commented:
looks like a great start. Thanks.. i will try take a closer look on the weekend.

petergthatsme commented:
answers to your questions

> Initially, I tried to parametrize the tests with all the available esys or evals methods but realized that some combinations may not be suitable naturally. This is why the methods are hardcoded inside the functions.

yes, as a first pass that's fine i think. 

Although one thing I could imagine could be really nice, would be going through each element of `DIAG_METHODS.keys` and just running it with some fixed systems (e.g. `Fluxonium`)...  
To handle specific lib requirements (e.g. primme), one could could split things up into methods that would only consider a particular lib "backend"... so for example, one could have a test method that (1) checks that `primme` is installed (you're already doing that).. then (2) going through `DIAG_METHODS.keys`, picking off ones that contain "primme", and uses those for diagonalization... 
same could be done for other cases. 

>A quick question: if the custom diagonalization method is a string, should providing an eval method to the esys_method argument or vice versa raise an explicit error? Currently, the error reads ValueError: too many values to unpack (expected 2) when an evals method is passed to esys_method.

yes, that's a good idea; we could check for for that and print a nice error to the user.

Yash-10 commented:
Great, thank you for replying! I'll try those ideas this weekend.

Yash-10 commented:
> Although one thing I could imagine could be really nice, would be going through each element of DIAG_METHODS.keys and just running it with some fixed systems (e.g. Fluxonium)...
To handle specific lib requirements (e.g. primme), one could could split things up into methods that would only consider a particular lib "backend"... so for example, one could have a test function that (1) checks that primme is installed (you're already doing that).. then (2) goes through DIAG_METHODS.keys, and picks off ones that contain "primme", which it then uses for diagonalization...
Same could be done for other "backends".

This looks like a great idea. I've tried this, but it looks like some methods don't agree with the default.

```py
def test_custom_diagonalization_evals_method_matches_default_primme():
    library = 'scipy'
    try:
        importlib.import_module(library)
    except ImportError:
        warnings.warn(f'Package {library} not installed; skipping test', ImportWarning)
        return

    primme_diag_methods = [
        method for method in diag_methods if method.split('_')[1] == library and method.split('_')[0] == 'evals'
    ]

    for method in primme_diag_methods:
        fluxonium = Fluxonium(
            EJ=8.9, EC=2.5, EL=0.5, flux = 0.5, cutoff = 120,
            evals_method=method
        )

        evals = fluxonium.eigenvals()

        fluxonium.evals_method = None

        evals_default = fluxonium.eigenvals()

        print(method)
        print(evals)
        print(evals_default)
```

I get:
```bash
evals_scipy_dense
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
evals_scipy_sparse
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
evals_scipy_sparse_SM
[-0.26659009 -0.63001146  8.75633751 11.69881054 16.67252211 17.42908833]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
evals_scipy_sparse_LA_shift-inverse
[22.52228391 18.66447202 17.42908833 16.67252211 11.69881054  8.75633751]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
evals_scipy_sparse_LM_shift-inverse
[17.42908833 16.67252211 11.69881054  8.75633751 -0.26659009 -0.63001146]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
```

I understand that the `_SM` methods have values ordered by magnitude, so one can compare it with the default by appropriate sorting. However, `LA_shift-inverse` and `LM_shift-inverse` are drastically different.

With `primme`, I get this:

```bash
evals_primme_sparse
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
evals_primme_sparse_SM
[-0.26659009 -0.63001146  8.75633751 11.69881054 16.67252211 17.42908833]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
evals_primme_sparse_LA_shift-inverse
[378.39084184 375.54643197 372.77911154 370.11569182 367.43665564
 364.54583971]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
evals_primme_sparse_LM_shift-inverse
[378.39084184 375.54643197 372.77911154 370.11569182 367.43665564
 364.54583971]
[-0.63001146 -0.26659009  8.75633751 11.69881054 16.67252211 17.42908833]
```

Do you agree this is unexpected @petergthatsme?

petergthatsme commented:
Yes, there are subtleties with how these can be used (i do add a warning in the docs, but maybe should stress this point even mre); we can indeed expect the magnitude-based methods to fail if the spectrum has negative energies. One quick-hack way out of the problem in this case would be to add a constant EJ to the Hamiltonian, which should enforce we have no negative eigenvalues. 

Regarding the shift-invert/LA based methods; maybe for now do not include them in the tests. I will at some stage take a closer look to see what could be going wrong there.

Thanks for putting in the effort into this issue - your work will be useful to us.

Yash-10 commented on scqubits/tests/diag_tests.py:
```diff
@@ -0,0 +1,281 @@
+        method for method in diag_methods
+        if method.split('_')[1] == library
+        and method.split('_')[0] == ('evals' if evals else 'esys')
+        and '_shift-inverse' not in method
```
`_shift-inverse` type methods are omitted from all tests since it seems they need to be investigated in more detail to see why do they not agree with the default method. I'll try to see if I can find the reason.

Yash-10 commented on scqubits/tests/diag_tests.py:
```diff
@@ -0,0 +1,281 @@
+        if method.split('_')[1] == library
+        and method.split('_')[0] == ('evals' if evals else 'esys')
+        and '_shift-inverse' not in method
+        and '_SM' not in method  # TODO: this is temporary until we find a fix for adding EJ to the Hamiltonian.
```
For now, I have also omitted `_SM` type methods.

For your below suggestion:
> One quick-hack way out of the problem in this case would be to add a constant EJ to the Hamiltonian, which should enforce we have no negative eigenvalues.

That sounds interesting to try and, in principle, physically equivalent. I can try changing the source code of the Qubit class, for example.

The other hacky alternative I had was to allow all eigenvalues instead (using `evals_count`) and also eigenvectors, so that if we sort the output of `_SM`, we should get the same values as the default, I think.

petergthatsme replied:
thanks.. yes, that's a good move "for now"...  adding the offset EJ to the main code, might confuse people, so we'd probably have to do it carefully... given you set up the core code with this PR, it will be easy to play and modify later what tests we re doing and how. 
i will take closer look at your code in a day or two.. and just merge once all ok.

Yash-10 replied:
Great, thanks! So I'll refrain from changing anything in the main code here. I've added a few comments inside the tests to clarify why we are performing some "hacks" until the necessary bugs are fixed, which might be helpful later.

Yash-10 commented:
@petergthatsme we are close to the end of the hackathon, so could you take a look at this? I don't think there was anything significant left to work on unless you have any comments.

petergthatsme commented:
thanks for making the updates. one last thing: could you please change `diag_tests.py` to `test_diag.py` (that's my bad - sorry!.. i know in the instructions i explicitly wrote the former, whereas i meant to write the latter... sigh.). 
I will merge after that change. 
Thanks again for your efforts.

Yash-10 commented:
Thank you, @petergthatsme; the file has been renamed.
